sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java

```diff
@@ -68,58 +70,109 @@
    * }</pre>
    */
   public static <T> WithTimestamps<T> of(SerializableFunction<T, Instant> fn) {
-    return new WithTimestamps<>(fn, Duration.ZERO);
+    return new WithTimestamps<>(fn, null, null);
   }
 
   ///////////////////////////////////////////////////////////////////
 
   private final SerializableFunction<T, Instant> fn;
-  private final Duration allowedTimestampSkew;
+  // If allowedTimestampSkew is not null, use the deprecated behavior which emits data as late.
+  private final @Nullable Duration allowedTimestampSkew;
+  // If watermarkDelay is not null, emit data within the watermark delay bound as on time.
+  private final @Nullable Duration watermarkDelay;
 
-  private WithTimestamps(SerializableFunction<T, Instant> fn, Duration allowedTimestampSkew) {
+  private WithTimestamps(
+      SerializableFunction<T, Instant> fn,
+      @Nullable Duration allowedTimestampSkew,
+      @Nullable Duration watermarkDelay) {
+    checkArgument(
+        allowedTimestampSkew == null || watermarkDelay == null, "Both delays cannot be non-null.");
```

Comment:
should this be a single constructor with a required null field, or should this be separated out?

Reply:
its private, I could write 3 constructors one for each use case where each is called once, but that feels like overkill

Reply:
I suppose, but it is strange to have a method where a parameter must be null

Reply:
Tried to do this- then I realized that its infeasible, because the two parameters have the same type. Closing as infeasible since you couldn't form the overload set.